### PR TITLE
stdenv/generic/setup.sh: enable parallel installs for parallel builds

### PR DESCRIPTION
**Pull request by trofi**
The primary motivating example is openssl:

Before the change full package build took 1m54s minutes. After the change full package build takes 59s.

About a 2x speedup.

The difference is visible because openssl builds hundreds of manpages spawning a perl process per manual in `install` phase. Such a workload is very easy to parallelize.

Another example would be `autotools`+`libtool` based build system where install step requires relinking. The more binaries there are to relink the more gain it will be to do it in parallel.

The change enables parallel installs by default only for buiilds that already have parallel builds enabled. There is a high chance those build systems already handle parallelism well but some packages will fail.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/stdenv/generic/setup.sh [hunk 1]
```diff
@@ -1337,6 +1337,7 @@ installPhase() {
 
     # shellcheck disable=SC2086
     local flagsArray=(
+        ${enableParallelInstalling:+-j${NIX_BUILD_CORES}}
```
Comment (alyssais) [inline pkgs/stdenv/generic/setup.sh:1340]: Are there other build systems we should be supporting this in?  Ninja, for example?

Comment (trofi) [reply pkgs/stdenv/generic/setup.sh:1340]: Yeah, extending it to other build systems sounds reasonable.

Comment (trofi) [reply pkgs/stdenv/generic/setup.sh:1340]: Switched to `draft` for now. Will try to extend it to (or at least audit for relevance):

- pkgs/development/tools/build-managers/bmake/setup-hook.sh
- pkgs/development/tools/build-managers/cmake/setup-hook.sh
- pkgs/development/tools/build-managers/ninja/setup-hook.sh
- pkgs/development/tools/build-managers/scons/setup-hook.sh

Comment (trofi) [reply pkgs/stdenv/generic/setup.sh:1340]: Attempted to cover:
    - cmake (enabled by default, similar to builds)
    - ninja (set parallelism explicitly, don't rely on default)
    - bmake (enable when requested)
    - scons (enable when requested)
    - meson (set parallelism explicitly, don't rely on default)
    - waf (set parallelism explicitly, don't rely on default)
    - qmake-4/5/6 (enable by default, similar to builds)
    - xorg (always enable, similar to builds)

**Comment by Atemu** [comment]
If a package build is doing anything other than installing files in the installPhase, that smells to me that it's building some previously unbuilt target that should have been built in the buildPhase but wasn't. Your example might be still *building* the docs as a dependency of the install.

The correct fix in that case would be to explicitly build all targets that are to be installed rather than accelerating builds in the installPhase.

Now of course there might still be a benefit to installing in parallel (I would expect there to be one) but that should be evaluated independently of the coincidental fix to the potential issue above.

**Comment by trofi** [comment]
> If a package build is doing anything other than installing files in the installPhase, that smells to me that it's building some previously unbuilt target that should have been built in the buildPhase but wasn't. Your example might be _building_ the docs as a dependency of the install.
> 
> The correct fix in that case would be to explicitly build all targets that are to be installed rather than accelerating builds in the installPhase.

I will not stop you from fixing it. But I do not think such fixes belong to `nixpkgs` and should be done upstream. I would say it's a good enough workaround to just make these cases faster. No need to penalize `nixpkgs` users for upstream faults. Calling `make` with and without `-j` is inconsistent. I would say `openssl` is a perfect example where it helps.

> Now of course there might still be a benefit to installing in parallel (I would expect there to be one) but that should be evaluated independently of the coincidental fix to the potential issue above.

`ghc` is another example of heavyweight `make install`. `gcc` is yet another one. Both install many files and relink executables against target install directory.

**Comment by Atemu** [comment]
> I will not stop you from fixing it. But I do not think such fixes belong to `nixpkgs` and should be done upstream.

It really wouldn't be much of a fix; you'd merely declare the default + additional targets in `buildFlags`. Yes that is absolutely an upstream bug and they should fix it internally but that applies to many things we currently patch ourselves.

> I would say it's a good enough workaround to just make these cases faster. No need to penalize `nixpkgs` users for upstream faults. Calling `make` with and without `-j` is inconsistent. I would say `openssl` is a perfect example where it helps.

I generally agree but I fear doing so would mask the actual issue that should be worked around or at least reported.

Therefore, I'd advocate for this feature being handled the same as `enableParallelBuilding`; disabled by default unless `config.enableParallelInstallingByDefault` but enabled by default for select build systems that are known to handle it well and/or benefit from it.

Another unknown is whether this affects r13y (certain parallel builds have done so in the past) which is also why I'd prefer this being off by default.

> `ghc` is another example of heavyweight `make install`. `gcc` is yet another one. Both install many files and relink executables against target install directory.

That's great. My point is that this PR's merit should be measured by cases like that rather than (potentially, I haven't checked) broken build scripts.

**Comment by trofi** [comment]
> > I will not stop you from fixing it. But I do not think such fixes belong to `nixpkgs` and should be done upstream.
> 
> It really wouldn't be much of a fix; you'd merely declare the default + additional targets in `buildFlags`. Yes that is absolutely an upstream bug and they should fix it internally but that applies to many things we currently patch ourselves.

If we enable parallel installs by default you would not need to do even that. Which sounds like a nice decrease in maintenance burden.

> > I would say it's a good enough workaround to just make these cases faster. No need to penalize `nixpkgs` users for upstream faults. Calling `make` with and without `-j` is inconsistent. I would say `openssl` is a perfect example where it helps.
> 
> I generally agree but I fear doing so would mask the actual issue that should be worked around or at least reported.

I don't think we have any signal today that we will lose by enabling install parallelism.

> Therefore, I'd advocate for this feature being handled the same as `enableParallelBuilding`; disabled by default unless `config.enableParallelInstallingByDefault` but enabled by default for select build systems that are known to handle it well and/or benefit from it.

Why is already used  `config.enableParallelInstallingByDefault` not enough? If the package can handle build parallelism there should be a high chance of handling install parallelism at least in most cases.

> Another unknown is whether this affects r13y (certain parallel builds have done so in the past) which is also why I'd prefer this being off by default.

Why is it relevant here? If parallel builds break reproducibility then we fix the reproducibility issue one way or another. Why mask it? Note that it is not as drastic as as breaking package build. I would expect packages to fail the build be a more common issue that introduce output non-determinism.

> > `ghc` is another example of heavyweight `make install`. `gcc` is yet another one. Both install many files and relink executables against target install directory.
> 
> That's great. My point is that this PR's merit should be measured by cases like that rather than (potentially, I haven't checked) broken build scripts.

I don't see a big difference in those. Both can do more or do less in `make install` as long as it produces needed outcome.

**Comment by amjoseph-nixpkgs** [comment]
> I generally agree but I fear doing so would mask the actual issue that should be worked around or at least reported.

The "actual issue" is already masked.

There is no `assert` or build failure if you are compiling things during the `installPhase`.  Serializing the `installPhase` isn't an assertion or a check -- it's an annoyance, which we only notice in the really ridiculous cases like openssl.

If you want to unmask the issue, please write a PR that causes `cc-wrapper` to fail during the `installPhase`, and then a new `allowCompilationDuringInstallPhase` flag.  Then we would have a *real* check for the issue you're concerned about.


> Another unknown is whether this affects r13y (certain parallel builds have done so in the past) which is also why I'd prefer this being off by default.

This is a totally valid concern.  I think the proposal to do a few Hydra runs with this PR before merging it is a very good idea.

**Comment by Atemu** [comment]
> If we enable parallel installs by default you would not need to do even that. Which sounds like a nice decrease in maintenance burden.

That'd mean building things in the install phase and that just seems wrong to me.

> I don't think we have any signal today that we will lose by enabling install parallelism.

The signal is time and parallelism.

When refactoring ffmpeg, I mistakenly stopped building the general all target and only noticed because it was taking forever building ffmpeg on only one core despite `enableParallelBuilding`.

I wasn't building *anything* in the buildPhase there. I wouldn't have noticed that if the installPhase *built* the package in parallel.

> Why is already used `config.enableParallelInstallingByDefault` not enough? If the package can handle build parallelism there should be a high chance of handling install parallelism at least in most cases.

Ah, didn't notice it was guarded behind `enableParallelBuilding`. That looks sane to me.

> I don't see a big difference in those.

Not a big difference in what? Time?

I'm mostly concerned about weighing this PR's merit in the time improvements in regular builds rather than broken ones as the broken ones should be fixed rather than worked around.

> To unmask the issue would require a PR that causes `cc-wrapper` (and compilers for other languages) to fail during the `installPhase`, and add a `allowCompilationDuringInstallPhase` flag for the cases where upstream deliberately does things like running perl scripts during `make install`. Then we would have a _real_ check for the issue.

That sounds great.

**Comment by trofi** [comment]
> > I don't think we have any signal today that we will lose by enabling install parallelism.
> 
> The signal is time and parallelism.

Sounds a bit abstract. What query should I write against local logs or hydra to catch `openssl` and other packages like that?

f I were to build interesting the problematic cases in `nixpkgs` I would consider adding something queryable.

An example of a more actionable signal:
- CPU and wall-time seconds in build phase
- CPU and wall-time seconds in install time

That at least allows you to construct a query to find suspicious cases, like: "a lot (>10s) of time in install", "install time is a lot longer than build time" and similar.

I don't think making an install phase parallel by default masks something like that infeasible. And I don't think we have an equivalent of that today as a signal.

> When refactoring ffmpeg, I mistakenly stopped building the general all target and only noticed because it was taking forever building ffmpeg on only one core despite `enableParallelBuilding`.
> 
> I wasn't building _anything_ in the buildPhase there. I wouldn't have noticed that if the installPhase _built_ the package in parallel.

How did you detect that? Did you look at the timing reported in logs? Or happened to notice it interactively?

I agree this specific use case will look differently. I disagree it overweights the benefit of faster installs for everyone.

I would say that it's a feature to get faster install in this case: if the end result is the same then it's a cosmetic issue. If the result is different - you can find about it faster as the full build is still fast.

> > I don't see a big difference in those.
> 
> Not a big difference in what? Time?

Not a big difference in the type of the problem: both do something substantial in `install` phase that is worth running in parallel.

> I'm mostly concerned about weighing this PR's merit in the time improvements in regular builds rather than broken ones as the broken ones should be fixed rather than worked around.

I disagree that `openssl` build is broken: it's not ideal but it's not incorrect.

**Comment by 7c6f434c** [comment]
> Another unknown is whether this affects r13y (certain parallel builds have done so in the past) which is also why I'd prefer this being off by default.

Yes, if we don't always parallel-build due to race conditions making output less deterministic than it could be, and the main case where parallel install helps is building during the install phase, same considerations probably apply.

> Therefore, I'd advocate for this feature being handled the same as `enableParallelBuilding`; disabled by default unless `config.enableParallelInstallingByDefault` but enabled by default for select build systems that are known to handle it well and/or benefit from it.

Maybe it should just default to `enableParallelBuilding` then? And override in the very rare case where installation behaves _worse_ than main building.

Notably, `openssl` already has this flag set.

File: pkgs/stdenv/generic/setup.sh (hunk 1 above)
Comment (alyssais) [reply pkgs/stdenv/generic/setup.sh:1340]: Thanks!

**Review by Mindavi — APPROVED** [review]
I think this is a good idea and making builds faster helps everyone. A jobset would be a good idea to hash out other big offenders, as mentioned earlier.

File: pkgs/servers/x11/xorg/builder.sh [hunk 2]
```diff
@@ -37,5 +37,6 @@ fi
```
Comment (SuperSandro2000) [inline pkgs/servers/x11/xorg/builder.sh:37]: totally unrelated to this PR but I wouldn't mind if someone could remove this legacy builder script and put the content into mkDerivation like normally.

**Review by SuperSandro2000 — APPROVED** [review]

**Comment by trofi** [comment]
Rebased against today's merge base of `master` and `staging` in preparation for hydra jobset. No changes otherwise.

**Comment by vcunat** [comment]
Jobset: https://hydra.nixos.org/jobset/nixpkgs/pr-217568-stdenv-parallel-install

**Review by SuperSandro2000 — APPROVED** [review]

**Review by 7c6f434c — APPROVED** [review]

**Comment by trofi** [comment]
Hydra run is complete \o/: https://hydra.nixos.org/eval/1791763?compare=trunk&full=1#tabs-now-fail

It uncovered only 12 packages that fail to install (all have a workaround now). Does not sound too bad.

**Review by mweinelt — APPROVED** [review]

**Comment by trofi** [comment]
A bit of fallout in `staging` post merge:
- [ ] ruby: https://github.com/NixOS/nixpkgs/pull/221442

**Comment by trofi** [comment]
From `staging-next` failures:

- [ ] _389-ds-base: https://github.com/NixOS/nixpkgs/pull/222161

**Comment by vcunat** [comment]
I fixed more failures that seemed caused by this.  I suppose the best reference generally is
```bash
git grep 'enableParallelInstalling = false;'
```
or
```bash
git log -S enableParallelInstalling
```

**Pull request merged**